### DISCUSSION TITLE
Regarding multiple key share entries in ClientHello (TLS 1.3)

### DISCUSSION THREAD

**Discussion by RokadeAditi** [openssl/openssl#21601] [Q&A]
Hi

I am checking if we can  support Key_Share extension requirement using OpenSSL 3.1.1. in ClientHello as per RFC 8446.
As per RFC 8446 :
"Clients can offer as many KeyShareEntry values as the number of
   supported groups it is offering, each representing a single set of
   key exchange parameters."

Do we send KeyShareEntry for each group added in "supported_group" ? In short does OpenSSL support sending multiple key shares in ClientHello (TLS 1.3)?

Can someone please help. 
Appreciate your help here.

Thanks.

**Comment by RokadeAditi** [comment]
When I set 2 groups, I can see key share sent only for the first group from the preferred list. 

eg. I have 2 groups as given below : 
Supported Groups (2 groups)
    Supported Group: secp521r1 (0x0019)
    Supported Group: x448 (0x001e)

Then I can see key share for only first group in wireshark log --->

Key Share Entry: Group: secp521r1, Key Exchange length: 133
    Group: secp521r1 (25)
    Key Exchange Length: 133

Can we configure openssl to include Key Share Entry for both the groups?

**Comment by RokadeAditi** [comment]
@mattcaswell Can you please help me on this?

**Comment by mattcaswell** [comment]
We don't support this. OpenSSL only supports sending a single key share.

**Reply by RokadeAditi** [reply]
Oh ok.  Thanks for the response.
Is there any plan to support this in future?

**Reply by mattcaswell** [reply]
No one is currently working on it and there are no specific plans to add it. OTOH, if someone created a PR to add that feature it would be considered.

**Reply by RokadeAditi** [reply]
Ok. Thank you.
Can I raise PR for it then? Because we need that support as per RFC 8446.

**Reply by mattcaswell** [reply]
Please feel free to raise a PR and we can take a look at it. It wouldn't be backported to stable releases (3.0/3.1) but it could be added to a forthcoming release (3.2 or 3.3).

**Reply by RokadeAditi** [reply]
Yes sure. Thank you.
I have raised this issue for feature request :
https://github.com/openssl/openssl/issues/21633

**Reply by RokadeAditi** [reply]
Hi @mattcaswell,

When can we expect forthcoming release of OpenSSL 3.2?

**Reply by mattcaswell** [reply]
Tentatively targeted for October, but feature freeze will be before that.
